text/0020-unique-record-keys.md

```diff
@@ -0,0 +1,103 @@
+# Disallow duplicate keys in Cedar records
+
+## Related issues and PRs
+
+- Reference Issues: related to [cedar#169](https://github.com/cedar-policy/cedar/pull/169)
+- Implementation PR(s):
+
+## Timeline
+
+- Start Date: 2023-07-14
+- Date Entered FCP:
+- Date Accepted:
+- Date Landed:
+
+## Summary
+
+Today, Cedar allows duplicate keys in record literals (and other record values, including `context`), with last-value-wins semantics.
+This RFC proposes to disallow duplicate keys in record values, making that an error.
+
+## Basic example
+
+Today,
+```
+permit(principal, action, resource)
+when {
+    resource.data == { foo: 2, foo: 3 }
+};
+```
+is a valid Cedar policy. The `foo: 2` is dropped and `foo` will have the value `3` in the record literal.
+In other words, in Cedar,
+```
+{ foo: 2, foo: 3 } == { foo: 3 }
+```
+evaluates to `true`.
+
+More precisely, Cedar's semantics does not just drop values prior to the last, but fully evaluates them before throwing them away. This has consequences in the presence of errors. For instance,
+```
+{ foo: 1 + "hello", foo: 3 }
+```
+does not evaluate to `{ foo: 3 }` as you might expect based on the previous example, but rather, it results in an evaluation error.
+
+This RFC proposes that all of the above expressions and policies would result in an evaluation error due to duplicate keys in a record.
+
+## Motivation
+
+There are two main motivations for this change.
+
+First, in the status quo, Cedar policy ASTs actually cannot losslessly roundtrip to EST and back.
```

Comment:
Is it possible to let EST to contain lists instead of maps?

Reply:
Yes; I discuss this under Alternatives

Reply:
My bad. Thank you.